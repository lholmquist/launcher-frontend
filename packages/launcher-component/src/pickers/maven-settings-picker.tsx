import React from 'react';
import { optionalBool } from '../core/stuff';
import { LaunchTextInput } from '../core/text-input/text-input';
import { TogglePanel } from '../core/toggle-panel/toggle-panel';
import { InputProps, Picker } from '../core/types';

export interface MavenSettingsPickerValue {
  groupId?: string;
  artifactId?: string;
  version?: string;
  name?: string;
  description?: string;
  packageName?: string;
}

interface MavenSettingsPickerProps extends InputProps<MavenSettingsPickerValue> {
  showVersion?: boolean;
  showMoreOptions?: boolean;
}

const VALUE_REGEXP = /^[a-z][a-z0-9-\.]{3,63}$/;

export const MavenSettingsPicker: Picker<MavenSettingsPickerProps, MavenSettingsPickerValue> = {
  checkCompletion: value => !!value.groupId && VALUE_REGEXP.test(value.groupId)
    && !!value.artifactId && VALUE_REGEXP.test(value.artifactId) && !!value.version,
  Element: props => {
    const isValid = (value?: string) => !!value && VALUE_REGEXP.test(value || '');
    return (
      <div className="pf-c-form">
        <LaunchTextInput
          label="Group"
          isRequired
          helperTextInvalid="Please provide a valid groupId"
          type="text"
          id="groupId"
          name="groupId"
          aria-label="Maven groupId name"
          value={props.value.groupId || ''}
          onChange={value => props.onChange({ ...props.value, groupId: value })}
          pattern={VALUE_REGEXP.source}
          isValid={isValid(props.value.groupId)}
        />
        <LaunchTextInput
          label="Artifact"
          isRequired
          helperTextInvalid="Please provide a valid artifactId"
          type="text"
          id="artifactId"
          name="artifactId"
          aria-label="Maven artifactId name"
          value={props.value.artifactId || ''}
          onChange={value => props.onChange({ ...props.value, artifactId: value })}
          pattern={VALUE_REGEXP.source}
          isValid={isValid(props.value.artifactId)}
        />
        {optionalBool(props.showVersion, false) && (
          <LaunchTextInput
            label="Version"
            helperTextInvalid="Please provide a version number"
            isRequired
            type="text"
            id="version"
            name="version"
            aria-label="Maven version number"
            value={props.value.version || ''}
            onChange={value => props.onChange({ ...props.value, version: value })}
            isValid={!!props.value.version}
          />
        )}
        {optionalBool(props.showMoreOptions, true) && (
          <TogglePanel title="Options">
            <div className="pf-c-form">
              <LaunchTextInput
                label="Name"
                helperTextInvalid="Please provide a name"
                isRequired
                type="text"
                id="name"
                name="name"
                aria-label="Maven name"
                value={props.value.name || ''}
                onChange={value => props.onChange({ ...props.value, name: value })}
                isValid={true}
              />
              <LaunchTextInput
                label="Description"
                helperTextInvalid="Please provide a description"
                isRequired
                type="text"
                id="description"
                name="description"
                aria-label="Maven description"
                value={props.value.description || ''}
                onChange={value => props.onChange({ ...props.value, description: value })}
                isValid={true}
              />
              <LaunchTextInput
                label="Package Name"
                helperTextInvalid="Please provide a package name"
                isRequired
                type="text"
                id="packageName"
                name="packageName"
                aria-label="Maven package name"
                value={props.value.packageName || ''}
                onChange={value => props.onChange({ ...props.value, packageName: value })}
                isValid={true}
              />
            </div>
          </TogglePanel>
        )}
      </div>
    );
  }
};
