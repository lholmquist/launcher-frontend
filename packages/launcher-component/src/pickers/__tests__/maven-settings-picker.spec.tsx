import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import { FormPanel } from '../../core/form-panel/form-panel';

import { MavenSettingsPicker } from '../maven-settings-picker';

describe('<MavenSettingsPicker />', () => {
  it('renders the MavenSettingsPicker correctly', () => {
    const comp = render(<MavenSettingsPicker.Element value={{groupId: '', version: '', artifactId: ''}} onChange={() => {}}/>);
    expect(comp.asFragment()).toMatchSnapshot();
  });

  it('show error for invalid data', () => {
    const handleSave = jest.fn();
    const comp = render(
      <FormPanel
        initialValue={{}}
        validator={MavenSettingsPicker.checkCompletion}
        onSave={handleSave}
        onCancel={() => {}}
      >
        {(inputProps) => (<MavenSettingsPicker.Element {...inputProps} showVersion />)}
      </FormPanel>
    );

    const groupIdField = comp.getByLabelText('Maven groupId name');
    fireEvent.change(groupIdField, { target: { value: 'invalid name' } });
    const artifactIdField = comp.getByLabelText('Maven artifactId name');
    fireEvent.change(artifactIdField, { target: { value: 'invalid name' } });
    const versionField = comp.getByLabelText('Maven version number');
    fireEvent.change(versionField, { target: { value: '1' } });
    expect(comp.asFragment()).toMatchSnapshot();
  });
});
