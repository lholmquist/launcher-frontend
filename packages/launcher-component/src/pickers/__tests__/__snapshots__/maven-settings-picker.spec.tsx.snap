// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MavenSettingsPicker /> renders the MavenSettingsPicker correctly 1`] = `
<DocumentFragment>
  <div
    class="pf-c-form"
  >
    <div
      class="pf-c-form__group"
    >
      <label
        class="pf-c-form__label"
        for="groupId"
      >
        Group
      </label>
      <input
        aria-invalid="false"
        aria-label="Maven groupId name"
        class="pf-c-form-control"
        id="groupId"
        name="groupId"
        pattern="^[a-z][a-z0-9-\\\\.]{3,63}$"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="pf-c-form__group"
    >
      <label
        class="pf-c-form__label"
        for="artifactId"
      >
        Artifact
      </label>
      <input
        aria-invalid="false"
        aria-label="Maven artifactId name"
        class="pf-c-form-control"
        id="artifactId"
        name="artifactId"
        pattern="^[a-z][a-z0-9-\\\\.]{3,63}$"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="panel "
    >
      <div
        class="pf-c-form"
      >
        <div
          class="pf-c-form__group"
        >
          <label
            class="pf-c-form__label"
            for="name"
          >
            Name
          </label>
          <input
            aria-invalid="false"
            aria-label="Maven name"
            class="pf-c-form-control"
            id="name"
            name="name"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="pf-c-form__group"
        >
          <label
            class="pf-c-form__label"
            for="description"
          >
            Description
          </label>
          <input
            aria-invalid="false"
            aria-label="Maven description"
            class="pf-c-form-control"
            id="description"
            name="description"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="pf-c-form__group"
        >
          <label
            class="pf-c-form__label"
            for="packageName"
          >
            Package Name
          </label>
          <input
            aria-invalid="false"
            aria-label="Maven package name"
            class="pf-c-form-control"
            id="packageName"
            name="packageName"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <a
      aria-disabled="false"
      aria-label="Expand panel"
      class="pf-c-button pf-m-link"
    >
      <span>
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 448 512"
          width="1em"
        >
          <path
            d="M448 80v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V80c0-26.5 21.5-48 48-48h352c26.5 0 48 21.5 48 48zM92.5 220.5l123 123c4.7 4.7 12.3 4.7 17 0l123-123c7.6-7.6 2.2-20.5-8.5-20.5H101c-10.7 0-16.1 12.9-8.5 20.5z"
            transform=""
          />
        </svg>
         More
      </span>
       Options
    </a>
  </div>
</DocumentFragment>
`;

exports[`<MavenSettingsPicker /> show error for invalid data 1`] = `
<DocumentFragment>
  <div
    class="form-panel"
    style="padding: 20px;"
  >
    <div
      class="pf-c-form"
    >
      <div
        class="pf-c-form__group"
      >
        <label
          class="pf-c-form__label"
          for="groupId"
        >
          Group
        </label>
        <input
          aria-invalid="false"
          aria-label="Maven groupId name"
          class="pf-c-form-control"
          id="groupId"
          name="groupId"
          pattern="^[a-z][a-z0-9-\\\\.]{3,63}$"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="pf-c-form__group"
      >
        <label
          class="pf-c-form__label"
          for="artifactId"
        >
          Artifact
        </label>
        <input
          aria-invalid="false"
          aria-label="Maven artifactId name"
          class="pf-c-form-control"
          id="artifactId"
          name="artifactId"
          pattern="^[a-z][a-z0-9-\\\\.]{3,63}$"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="pf-c-form__group"
      >
        <label
          class="pf-c-form__label"
          for="version"
        >
          Version
        </label>
        <input
          aria-invalid="false"
          aria-label="Maven version number"
          class="pf-c-form-control"
          id="version"
          name="version"
          required=""
          type="text"
          value="1"
        />
      </div>
      <div
        class="panel "
      >
        <div
          class="pf-c-form"
        >
          <div
            class="pf-c-form__group"
          >
            <label
              class="pf-c-form__label"
              for="name"
            >
              Name
            </label>
            <input
              aria-invalid="false"
              aria-label="Maven name"
              class="pf-c-form-control"
              id="name"
              name="name"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="pf-c-form__group"
          >
            <label
              class="pf-c-form__label"
              for="description"
            >
              Description
            </label>
            <input
              aria-invalid="false"
              aria-label="Maven description"
              class="pf-c-form-control"
              id="description"
              name="description"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="pf-c-form__group"
          >
            <label
              class="pf-c-form__label"
              for="packageName"
            >
              Package Name
            </label>
            <input
              aria-invalid="false"
              aria-label="Maven package name"
              class="pf-c-form-control"
              id="packageName"
              name="packageName"
              required=""
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <a
        aria-disabled="false"
        aria-label="Expand panel"
        class="pf-c-button pf-m-link"
      >
        <span>
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 448 512"
            width="1em"
          >
            <path
              d="M448 80v352c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V80c0-26.5 21.5-48 48-48h352c26.5 0 48 21.5 48 48zM92.5 220.5l123 123c4.7 4.7 12.3 4.7 17 0l123-123c7.6-7.6 2.2-20.5-8.5-20.5H101c-10.7 0-16.1 12.9-8.5 20.5z"
              transform=""
            />
          </svg>
           More
        </span>
         Options
      </a>
    </div>
    <hr
      class="separator"
    />
    <div
      class="pf-l-toolbar"
    >
      <div
        class="pf-l-toolbar__group"
      >
        <button
          aria-label="Save form"
          class="pf-c-button pf-m-primary pf-m-disabled"
          disabled=""
          type="button"
        >
          Save
        </button>
      </div>
      <div
        class="pf-l-toolbar__group"
      >
        <button
          aria-label="Cancel form"
          class="pf-c-button pf-m-secondary"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
